### PR TITLE
fix(http): disable HTTP idle timeout by default + tests/docs (follow-up to #179)

### PR DESCRIPTION
**Pull request by jmrplens**
## Description

Follow-up to #179 (thanks @turchinc for the original flag and the diagnosis). That PR added `--http-idle-timeout` to stop the hardcoded `http.Server` timeouts from severing long-lived MCP **Streamable HTTP** (SSE) sessions. This PR rounds out the work and changes the default so the fix applies out of the box.

### Why change the default?

#179 kept the default at `120s` — i.e. the exact value that was reported as too aggressive. With default flags, operators still lose long-lived sessions unless they discover and set the flag. The expected behavior is that connections are **not** recycled prematurely unless an operator opts into a low value. So the default is now `0` (idle closure disabled), making `--session-timeout` (30m) the effective idle lifetime.

## Changes

**Code (`cmd/server/main.go`)**
- Default `--http-idle-timeout` `120s` → `0` (disabled); positive values still honored.
- Named, documented timeout constants (no magic numbers); extracted `newHTTPServer` so the timeout policy is unit-testable without binding a port.
- `WriteTimeout` is coupled to at least the effective idle timeout. This matters because the go-sdk SSE writer (`mcp/event.go`, `mcp/streamable.go`) only flushes and never resets the write deadline, so `WriteTimeout` would otherwise bound the entire SSE stream lifetime.
- Help text / flag usage reworded for `0 = disabled`.

**Tests (`cmd/server/main_test.go`)**
- `TestEffectiveIdleTimeout`, `TestNewHTTPServer_Timeouts` (idle/write/read matrix), `TestRunHTTP_NegativeIdleTimeout_Rejected`, `TestDefaultHTTPIdleTimeout_DisabledByDefault`, plus a help-output assertion for the flag.

**Docs**
- Markdown: `cli-reference`, `http-server-mode` (new timeout-model section + troubleshooting row), `configuration`, `troubleshooting` (corrected misleading "increase --session-timeout" advice), `architecture`, `env-reference`, `CLAUDE.md`.
- Astro site EN + ES: `http-server`, `configuration`, `troubleshooting`.
- Regenerated `llms.txt` / `llms-full.txt` and `docs/testing/testing.md`.

**Version**
- Bump `2.2.0` → `2.2.1` (`VERSION`, `CLAUDE.md`). Release manifests (`server.json`, `.plugin/plugin.json`) are intentionally left to the release flow (`scripts/update-server-json-sha.sh`).

## How to test

```bash
gitlab-mcp-server --http                      # idle closure disabled by default
gitlab-mcp-server --http --http-idle-timeout=30m   # explicit value honored
gitlab-mcp-server --http --http-idle-timeout=-1s   # rejected with an error
```

## Verification

`go build ./...`, `go test ./cmd/server/`, `make analyze` (golangci-lint config + fmt + run, govulncheck, markdownlint), `format_md_tables --check`, and `gen_testing_docs --check` all pass.

## Breaking changes

Default behavior changes: the HTTP server no longer closes idle connections by default (was `120s`). Long-lived MCP sessions now persist until `--session-timeout` or client disconnect. Operators who want aggressive idle recycling can set a positive `--http-idle-timeout`. Behind a reverse proxy, the proxy's own read/idle timeout becomes the limiting factor.

## Summary by Sourcery

Change HTTP server idle timeout to be disabled by default and align server timeouts, tests, and docs to ensure long-lived Streamable HTTP sessions are not prematurely closed.

Bug Fixes:
- Prevent long-lived MCP Streamable HTTP (SSE) sessions from being closed by default HTTP idle/write timeouts by disabling idle closure and coupling write timeouts appropriately.

Enhancements:
- Refactor HTTP server construction into a reusable helper with named timeout constants and an explicit idle-timeout policy.
- Validate and reject negative --http-idle-timeout values at startup to avoid misconfiguration.

Documentation:
- Document the HTTP vs MCP timeout model, the default disabled HTTP idle timeout, and updated troubleshooting guidance in both English and Spanish HTTP server and configuration docs.
- Update CLI, configuration, environment reference, architecture, and troubleshooting documentation to reflect the new --http-idle-timeout behavior and defaults.
- Refresh generated testing documentation and LLM prompt asset files.

Tests:
- Add unit tests for effective idle timeout mapping, HTTP server timeout configuration, negative --http-idle-timeout rejection, default idle-timeout behavior, and help output coverage.

Chores:
- Bump the project version from 2.2.0 to 2.2.1 in version metadata and docs.

## Summary by CodeRabbit

* **New Features**
  * Added `--http-idle-timeout` flag to control HTTP server idle connection behavior (default: disabled, allowing `--session-timeout` to govern effective connection lifetime).

* **Documentation**
  * Updated configuration guides and troubleshooting documentation to explain HTTP idle timeout behavior and timeout model separation between HTTP layer and MCP session.

* **Tests**
  * Added test coverage for the new HTTP idle timeout flag.

* **Chores**
  * Bumped version to 2.2.1.

### CODE REVIEW

File: cmd/server/main.go [hunk 1]
```diff
@@ -855,16 +883,36 @@ func registerConfiguredToolSurfaceWithCatalog(server *mcp.Server, client *gitlab
...
+	idleTimeout := effectiveIdleTimeout(httpIdleTimeout)
+	return &http.Server{
+		Addr:              addr,
+		Handler:           handler,
+		ReadHeaderTimeout: baseHTTPReadHeaderTimeout,
+		ReadTimeout:       baseHTTPReadTimeout,
+		WriteTimeout:      max(baseHTTPWriteTimeout, idleTimeout),
+		IdleTimeout:       idleTimeout,
+	}
+}
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply cmd/server/main.go:936]: Good catch — fixed in 066c37d5. The global `WriteTimeout` is now kept at the fixed `baseHTTPWriteTimeout` (60s) as a slow-write/Slowloris guard for standard endpoints, and only the long-lived SSE GET stream disables its write deadline via `http.NewResponseController` in `sseWriteDeadlineMiddleware`. No more 10-year global write timeout.

File: cmd/server/main.go [hunk 2]
```diff
@@ -855,16 +883,36 @@ func registerConfiguredToolSurfaceWithCatalog(server *mcp.Server, client *gitlab
...
+// long-lived Streamable HTTP (SSE) responses are not severed before the idle
+// deadline; ReadHeaderTimeout/ReadTimeout are fixed request-read guards.
+func newHTTPServer(addr string, handler http.Handler, httpIdleTimeout time.Duration) *http.Server {
+	idleTimeout := effectiveIdleTimeout(httpIdleTimeout)
+	return &http.Server{
+		Addr:              addr,
+		Handler:           handler,
+		ReadHeaderTimeout: baseHTTPReadHeaderTimeout,
+		ReadTimeout:       baseHTTPReadTimeout,
+		WriteTimeout:      max(baseHTTPWriteTimeout, idleTimeout),
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply cmd/server/main.go]: Agreed and fixed in 066c37d5 — exactly as suggested: `WriteTimeout` stays at 60s and the SSE GET stream (`Accept: text/event-stream`) clears its write deadline via `http.NewResponseController`. Standard endpoints keep the guard.

File: cmd/server/main.go [hunk 3]
```diff
@@ -118,7 +118,36 @@ type httpConfig struct {
...
+
+	// baseHTTPReadTimeout bounds reading the full request. Streamable HTTP request
+	// bodies (JSON-RPC) are small, so this does not affect established SSE streams.
+	baseHTTPReadTimeout = 30 * time.Second
+
+	// baseHTTPWriteTimeout is the minimum response write timeout. It is raised to the
+	// effective idle timeout (see effectiveIdleTimeout) because the MCP go-sdk SSE
+	// writer never resets the write deadline, so WriteTimeout would otherwise bound
+	// the entire SSE stream lifetime and sever it prematurely.
+	baseHTTPWriteTimeout = 60 * time.Second
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply cmd/server/main.go]: Updated the `baseHTTPWriteTimeout` comment in 066c37d5 to reflect it is the fixed global guard and that SSE streams disable their deadline dynamically.

File: cmd/server/main_test.go [hunk 4]
```diff
@@ -3579,3 +3580,93 @@ func TestServeHTTP_ServerCardEndpoint_ReturnsToolList(t *testing.T) {
...
+			}
+			if srv.ReadHeaderTimeout != baseHTTPReadHeaderTimeout {
+				t.Errorf("ReadHeaderTimeout = %s, want %s", srv.ReadHeaderTimeout, baseHTTPReadHeaderTimeout)
+			}
+			if srv.ReadTimeout != baseHTTPReadTimeout {
+				t.Errorf("ReadTimeout = %s, want %s", srv.ReadTimeout, baseHTTPReadTimeout)
+			}
+		})
+	}
+}
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply cmd/server/main_test.go:3639]: Updated in 066c37d5: the test now asserts `WriteTimeout == baseHTTPWriteTimeout` for all idle values, and a new `TestSSEWriteDeadlineMiddleware_PassesThrough` covers SSE GET, plain GET, and MCP POST.

File: docs/http-server-mode.md [hunk 5]
```diff
@@ -380,6 +380,34 @@ If a client is idle for longer than `--session-timeout` (default: 30 minutes):
...
+stream that carries server-initiated notifications and 30s keep-alive pings. These
+are long-lived HTTP responses, so the `http.Server` `IdleTimeout` and `WriteTimeout`
+— not `--session-timeout` — determine whether they survive.
+
+Because the default `--http-idle-timeout` is `0` (disabled), the HTTP layer does not
+close idle connections, and `--session-timeout` is the effective idle lifetime. The
+response write timeout is automatically coupled to at least the idle timeout, so SSE
+streams are never severed before the idle deadline. If you set a low
+`--http-idle-timeout`, expect long-lived MCP sessions to drop when the connection
+idles past that value.
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply docs/http-server-mode.md]: Updated in 066c37d5 to describe the fixed 60s WriteTimeout plus per-SSE dynamic deadline disabling.

File: site/src/content/docs/operations/http-server.mdx [hunk 6]
```diff
@@ -216,6 +217,21 @@ graph TD
...
+### Timeout model: HTTP layer vs MCP session
+
+Two independent layers govern connection and session lifetime:
+
+- **MCP session** (`--session-timeout`, default `30m`): idle lifetime of the MCP session at the SDK transport level.
+- **HTTP idle connection** (`--http-idle-timeout`, default `0` = disabled): how long the underlying `http.Server` keeps an idle keep-alive connection before closing it.
+
+The server speaks the modern **Streamable HTTP** transport, which uses Server-Sent Events (`text/event-stream`) for streamed responses and the standalone stream that carries server-initiated notifications and keep-alive pings. Those are long-lived HTTP responses, so the HTTP `IdleTimeout` and `WriteTimeout` — not `--session-timeout` — decide whether they survive.
+
+Because `--http-idle-timeout` defaults to `0` (disabled), the HTTP layer does not close idle connections out of the box, and `--session-timeout` is the effective idle lifetime. The response write timeout is automatically coupled to at least the idle timeout, so SSE streams are never cut before the idle deadline. Set a low `--http-idle-timeout` only if you want idle connections recycled sooner.
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply site/src/content/docs/operations/http-server.mdx]: Updated in 066c37d5 (EN site).

File: site/src/content/docs/es/operations/http-server.mdx [hunk 7]
```diff
@@ -216,6 +217,21 @@ graph TD
...
+
+- **Sesión MCP** (`--session-timeout`, por defecto `30m`): vida útil de inactividad de la sesión MCP a nivel del transporte del SDK.
+- **Conexión HTTP inactiva** (`--http-idle-timeout`, por defecto `0` = desactivado): cuánto tiempo mantiene el `http.Server` una conexión keep-alive inactiva antes de cerrarla.
+
+El servidor habla el transporte moderno **Streamable HTTP**, que usa Server-Sent Events (`text/event-stream`) para las respuestas en streaming y el stream independiente que transporta las notificaciones iniciadas por el servidor y los pings de keep-alive. Esas son respuestas HTTP de larga duración, así que el `IdleTimeout` y `WriteTimeout` HTTP — no `--session-timeout` — deciden si sobreviven.
+
+Como `--http-idle-timeout` vale `0` (desactivado) por defecto, la capa HTTP no cierra las conexiones inactivas de fábrica, y `--session-timeout` es la vida efectiva de inactividad. El timeout de escritura de respuesta se acopla automáticamente a al menos el timeout de inactividad, de modo que los streams SSE nunca se cortan antes de la fecha límite de inactividad. Pon un `--http-idle-timeout` bajo solo si quieres reciclar antes las conexiones inactivas.
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply site/src/content/docs/es/operations/http-server.mdx]: Updated in 066c37d5 (ES site).

File: docs/http-server-mode.md [hunk 8]
```diff
@@ -380,6 +380,34 @@ If a client is idle for longer than `--session-timeout` (default: 30 minutes):
...
+| Layer                | Setting               | Default        | What it bounds                                                          |
+| -------------------- | --------------------- | -------------- | ----------------------------------------------------------------------- |
+| MCP session          | `--session-timeout`   | `30m`          | Idle lifetime of the MCP session (SDK transport level)                  |
+| HTTP idle connection | `--http-idle-timeout` | `0` (disabled) | TCP keep-alive idle time before the `http.Server` closes the connection |
+| HTTP response write  | derived               | ≥ idle timeout | Maximum write time for a response, raised to at least the idle timeout  |
+
+The server speaks the modern **Streamable HTTP** transport, which uses Server-Sent
+Events (`text/event-stream`) for streamed POST responses and the standalone GET
+stream that carries server-initiated notifications and 30s keep-alive pings. These
+are long-lived HTTP responses, so the `http.Server` `IdleTimeout` and `WriteTimeout`
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply docs/http-server-mode.md]: Fixed in 066c37d5: the docs now state that an active SSE response is bounded by `WriteTimeout`, not `IdleTimeout`.

File: docs/http-server-mode.md [hunk 9]
```diff
@@ -380,6 +380,34 @@ If a client is idle for longer than `--session-timeout` (default: 30 minutes):
...
 3. Next request from the same `(token, url)` pair creates a new MCP session on the existing server
 
+### Timeout model: HTTP layer vs MCP session
+
+Two independent layers govern how long a connection and a session live:
+
+| Layer                | Setting               | Default        | What it bounds                                                          |
+| -------------------- | --------------------- | -------------- | ----------------------------------------------------------------------- |
+| MCP session          | `--session-timeout`   | `30m`          | Idle lifetime of the MCP session (SDK transport level)                  |
+| HTTP idle connection | `--http-idle-timeout` | `0` (disabled) | TCP keep-alive idle time before the `http.Server` closes the connection |
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply docs/http-server-mode.md]: Fixed in 066c37d5: `IdleTimeout` is now described as the HTTP keep-alive wait for the next request, not a TCP timer.

File: VERSION [hunk 10]
```diff
@@ -1 +1 @@
-2.2.0
+2.2.1
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply VERSION]: Fixed in 066c37d5: removed the trailing newline from `VERSION` per `docs/development/development.md`.

[the remark this answers is not in the document]
Comment (jmrplens) [reply VERSION]: Fixed in 066c37d5: `VERSION` no longer has a trailing newline.

File: site/src/content/docs/es/operations/http-server.mdx [hunk 11]
```diff
@@ -51,6 +51,7 @@ El servidor comienza a escuchar en el puerto 8080 por defecto. El endpoint MCP e
 | `--embedded-resources`   | `true`                       | Incrustar URIs canónicas de recursos MCP en resultados de herramientas get\_\*                                                                         |
 | `--max-http-clients`     | `100`                        | Máximo de entradas únicas token+URL en el pool del servidor                                                                                            |
 | `--session-timeout`      | `30m`                        | Timeout de sesión MCP inactiva                                                                                                                         |
+| `--http-idle-timeout`    | `0` (desactivado)            | Timeout de conexión inactiva del servidor HTTP. `0` (por defecto) desactiva el cierre por inactividad, de modo que `--session-timeout` es la vida efectiva; usa una duración positiva para reciclar conexiones inactivas antes |
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply site/src/content/docs/es/operations/http-server.mdx:54]: Not applicable here — this file lives under `site/src/content/docs/es/`, which is the **intentional Spanish localization** of the Astro documentation site (mirrored from the English tree under `site/src/content/docs/`). The English-only project guideline targets development artifacts (source, comments, commit messages, tool metadata, ADRs, specs, etc.), not the translated user-facing docs site, which exists specifically to provide Spanish content. Translating these files to English would defeat the purpose of the `es/` locale. Resolving as won't-fix.

File: cmd/server/main.go [hunk 12]
```diff
@@ -855,16 +885,52 @@ func registerConfiguredToolSurfaceWithCatalog(server *mcp.Server, client *gitlab
...
 
+// sseWriteDeadlineMiddleware disables the per-connection write deadline for the
+// long-lived Server-Sent Events stream — the standalone GET stream that carries
+// MCP server-initiated notifications and keep-alive pings. The MCP go-sdk SSE
+// writer never resets the write deadline, so without this the server's
+// WriteTimeout would sever that stream. All other requests keep WriteTimeout as a
+// slow-write (Slowloris) guard, so disabling it globally is unnecessary and unsafe.
+func sseWriteDeadlineMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.Method == http.MethodGet && strings.Contains(r.Header.Get("Accept"), "text/event-stream") {
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply cmd/server/main.go]: Valid — fixed in 65d682a6. Confirmed the server runs Streamable HTTP without `JSONResponse`, so every POST response is `text/event-stream` and can stay open for the length of a tool call (e.g. awaiting an elicitation/sampling round-trip). The middleware now keys on `Accept: text/event-stream` for any method, so both the standalone GET stream and streamed POST responses keep an unbounded write deadline. `/health` and the unauthenticated server-card endpoint negotiate plain content and keep the 60s slow-write guard; only the authenticated `/mcp` endpoint skips it. Docs updated (EN + ES).
